src/main/java/dal/fatih/todoly/Main.java

```diff
@@ -43,42 +50,76 @@ public static Task handleCreateTask() {
     }
 
     public static void listAllTasks() {
-        if (tasks.isEmpty()) {
-            System.out.println("Task list is empty");
-        } else {
-            for (Map.Entry task : tasks.entrySet()) {
-                System.out.println(task.getValue());
-                System.out.println("----------------------------------");
+        try {
+            inputTask = new ObjectInputStream(new FileInputStream(file));
+            tasks = (HashMap) inputTask.readObject();

```

Comment:
Load tasks only once at the beginning of the program. Don't keep reading the file on every list and filter operation.

Reply:
Your tasks HashMap is already updated along with the file when there is an adding and a delete. Keep using tasks HashMap to list and filter. You can load the tasks from the file at the beginning of the main function before you get any input from the user.